syntax = 'proto3';
package com.science.foundry.tools.grpc_interfaces.vspin_with_loader;

message Command {
  oneof command {
    Home home = 1;
    CloseDoor close_door = 2;
    OpenDoor open_door = 3;
    LoadPlate load_plate = 4;
    UnloadPlate unload_plate = 5;
    Park park = 6;
    SpinCycle spin_cycle = 7;
    StopSpinCycle stop_spin_cycle = 8;
    ShowDiagsDialog show_diagnostics = 9;
  }

  message Home {}

  message CloseDoor {}

  message OpenDoor {
    int32 bucket_number = 1;  // 1 or 2
  }

  message LoadPlate {
    int32 bucket_number = 1;        // 1 or 2
    float gripper_offset = 2;       // 2.6 to 21.1
    float plate_height = 3;         // 0.0 to 48.0
    int32 speed = 4;               // 0 (slow) to 3 (fast)
    int32 options = 5;             // 0-7 bitfield options
  }

  message UnloadPlate {
    int32 bucket_number = 1;        // 1 or 2
    float gripper_offset = 2;       // 2.6 to 21.1
    float plate_height = 3;         // 0.0 to 48.0
    int32 speed = 4;               // 0 (slow) to 3 (fast)
    int32 options = 5;             // 0-7 bitfield options
  }

  message Park {}

  message SpinCycle {
    float velocity_percent = 1;
    float acceleration_percent = 2;
    float deceleration_percent = 3;
    int32 timer_mode = 4;
    int32 time = 5;
    int32 bucket_number_load = 6;      // 0 or 1 or 2
    int32 bucket_number_unload =7;      // 1 or 2
    float gripper_offset_load = 8;     // 2.6 to 21.1
    float gripper_offset_unload = 9;   // 2.6 to 21.1
    float plate_height_load = 10;       // 0.0 to 48.0
    float plate_height_unload = 11;    // 0.0 to 48.0
    int32 speed_load = 12;            // 0 (slow) to 3 (fast)
    int32 speed_unload = 13;          // 0 (slow) to 3 (fast)
    int32 load_options = 14;          // 0-7 bitfield options
    int32 unload_options = 15;        // 0-7 bitfield options
  }

  message StopSpinCycle {
    int32 bucket_number = 1;  // 1 or 2
  }

  message ShowDiagsDialog {
  }
}

message Config {
  string profile = 1;
}
